File: kora-java-validation/src/main/java/ru/tinkoff/kora/example/validation/ArgumentValidator.java
package ru.tinkoff.kora.example.validation;

import jakarta.annotation.Nullable;
import ru.tinkoff.kora.common.Component;
import ru.tinkoff.kora.common.annotation.Root;
import ru.tinkoff.kora.validation.common.annotation.*;

@Root
@Component
public class ArgumentValidator {

    @Valid
    public record User(@NotBlank String id,
                       @Size(min = 3, max = 6) String name,
                       @Nullable String status) {}

    @Validate
    public int calculate(@Valid User user,
                         @Range(from = 1, to = 900) int weight,
                         @Pattern("ME\\d+") String code) {
        return Integer.parseInt(code.substring(2));
    }
}

File: examples/kotlin/kora-kotlin-validation/src/main/kotlin/ru/tinkoff/kora/kotlin/example/validation/ArgumentValidator.kt
package ru.tinkoff.kora.kotlin.example.validation

import jakarta.annotation.Nullable
import ru.tinkoff.kora.common.Component
import ru.tinkoff.kora.common.annotation.Root
import ru.tinkoff.kora.validation.common.annotation.*

@Root
@Component
open class ArgumentValidator {
    @Validate
    open fun calculate(
        @Valid user: ArgumentUser,
        @Range(from = 1.0, to = 900.0) weight: Int,
        @Pattern("ME\\d+") code: String
    ): Int = code.substring(2).toInt()
}

@Valid
data class ArgumentUser(
    @field:NotBlank val id: String,
    @field:Size(min = 3, max = 6) val name: String,
    @field:Nullable val status: String?
)

